Implement BIP91 Reduced threshold Segwit MASF

jameshilliard opened the pull request:
This is an implementation of BIP91.
BIP91 details: https://github.com/bitcoin/bips/blob/master/bip-0091.mediawiki
Previous discussion:
https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2017-May/014380.html

shaolinfry committed:
Reduced threshold MASF to require segwit signalling

This is a BIP9 deployment at 65% that requires signalling
of segwit upon activation if segwit is not already
locked-in or activated. Signalling requirement stop at
segwit expiry, or activation.

jameshilliard committed:
Use version bit 4 with an 80% threshold for activation

jonasschnelli commented:
`bip68-112-113-p2p.py` are falling...

jameshilliard committed:
Add segsignal to getblockchaininfo

jameshilliard committed:
Use BIP91 activation threshold only on networks with a 95% BIP9 activation threshold.

jtimon commented on src/validation.cpp:
```diff
@@ -2918,6 +2930,13 @@ bool IsWitnessEnabled(const CBlockIndex* pindexPrev, const Consensus::Params& pa
 }
 
+// Check if Segregated Witness is Locked In
+bool IsWitnessLockedIn(const CBlockIndex* pindexPrev, const Consensus::Params& params)
```
This function could be static

jameshilliard replied:
Any reason IsWitnessEnabled isn't also static?

earonesty commented on src/versionbits.cpp:
```diff
@@ -148,7 +152,14 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
     int Period(const Consensus::Params& params) const { return params.nMinerConfirmationWindow; }
-    int Threshold(const Consensus::Params& params) const { return params.nRuleChangeActivationThreshold; }
+    int Threshold(const Consensus::Params& params) const {
+        if (params.nRuleChangeActivationThreshold == 1916 && params.nMinerConfirmationWindow == 2016 &&
```
can these magic numbers - when not being used as actual thresholds - be constants?

jtimon replied:
This commit seems unnecessary, just don't use Consensus::DEPLOYMENT_SEGSIGNAL on chains that shouldn't use it (like regtest and testnet3).

jameshilliard replied:
Fixed

jtimon commented on src/versionbits.cpp:
```diff
@@ -148,7 +152,13 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
+    int Threshold(const Consensus::Params& params) const {
+        if (params.vDeployments[id].bit == params.vDeployments[Consensus::DEPLOYMENT_SEGSIGNAL].bit &&
+            params.vDeployments[id].nStartTime == params.vDeployments[Consensus::DEPLOYMENT_SEGSIGNAL].nStartTime) {
+            return 1311; // 65% threshold for SEGSIGNAL only
```
Perhaps this can be a field on Consensus::Params? or at least a properly defined constant

jameshilliard replied:
This is now defined in chainparams

jtimon commented on src/versionbits.cpp:
```diff
@@ -148,7 +152,13 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
     int Period(const Consensus::Params& params) const { return params.nMinerConfirmationWindow; }
-    int Threshold(const Consensus::Params& params) const { return params.nRuleChangeActivationThreshold; }
+    int Threshold(const Consensus::Params& params) const {
+        if (params.vDeployments[id].bit == params.vDeployments[Consensus::DEPLOYMENT_SEGSIGNAL].bit &&
```
why not just ```if (id == Consensus::DEPLOYMENT_SEGSIGNAL)``` ?

jameshilliard replied:
Is the behavior for that identical?

jtimon commented on src/chainparams.cpp:
```diff
@@ -97,7 +97,7 @@ class CMainParams : public CChainParams {
 
         // Deployment of SEGSIGNAL
-        consensus.vDeployments[Consensus::DEPLOYMENT_SEGSIGNAL].bit = 2;
+        consensus.vDeployments[Consensus::DEPLOYMENT_SEGSIGNAL].bit = 4;
```
why bit 4 ? what's wrong with bit 2 ?

jameshilliard replied:
This was chosen because it was what was listed in the NYC agreement, I made a separate BIP that uses bit 2 however.

jtimon commented on src/rpc/blockchain.cpp:
```diff
@@ -1135,6 +1135,7 @@ UniValue getblockchaininfo(const JSONRPCRequest& request)
     softforks.push_back(SoftForkDesc("bip65", 4, tip, consensusParams));
     BIP9SoftForkDescPushBack(bip9_softforks, "csv", consensusParams, Consensus::DEPLOYMENT_CSV);
     BIP9SoftForkDescPushBack(bip9_softforks, "segwit", consensusParams, Consensus::DEPLOYMENT_SEGWIT);
+    BIP9SoftForkDescPushBack(bip9_softforks, "segsignal", consensusParams, Consensus::DEPLOYMENT_SEGSIGNAL);
```
I would squash this commit and the previous one.

jl2012 committed:
Override default confirmation window and threshold for SEGSIGNAL

jameshilliard committed:
Enforce BIP91 when locked in

jameshilliard commented:
I've [modified](https://github.com/bitcoin/bips/pull/547/files) the BIP91 spec slightly with a different threshold and immediate enforcement upon lock-in.

jl2012 committed:
Add regtest parameters for SEGSIGNAL

jl2012 committed:
Enforce SEGSIGNAL only when SEGWIT is in STARTED state

bit 1 signaling is not required (and has no meaning) unless the state is STARTED

jl2012 committed:
[qa] Add RPC test for SEGSIGNAL

jl2012 committed:
[qa] Fix tests under SEGSIGNAL

jl2012 committed:
Reduce the SEGSIGNAL signalling window to 336 blocks and enforce only when ACTIVE

jl2012 committed:
[qa] fix tests for SEGSIGNAL LOCKED_IN

jameshilliard commented:
I've [updated](https://github.com/bitcoin/bips/pull/548/files) BIP91 to use an even smaller confirmation window and changed it so enforcement isn't until activation in order to give a ~2 day grace period between lock-in and enforcement.

luke-jr commented:
NACK, this does nothing for users, and is unnecessary with BIP148 around the corner.

earonesty commented:
ACK: avoiding a chain split in a coordinated manner is certainly not "doing nothing for users".  Since the -bip148 option is not available, a BIP91 feature is the best alternative.

dooglus commented:
Apparently BIP91 was locked in today.

How can I make sure my node follows the correct chain? Is this pull request on a branch somewhere or do I need to merge it in my own local repository?

And what about the thousands of other nodes out there? Shouldn't this PR be merged to allow them to track the correct chain?

jameshilliard commented:
@dooglus BIP91 is mostly a miner enforced soft fork, you can use [this tree](https://github.com/segsignal/bitcoin).

dooglus commented:
As I understand Bitcoins miners don't enforce anything. They mine blocks which are either valid or invalid. It is the users who decide whether to accept their blocks or not, and thereby enforce the rules.

I don't see any argument here against merging this PR other than Luke's inaccurate complaint that BIP91 does nothing for users. What am I missing?

jameshilliard commented:
@dooglus Both miners and users have the ability to enforce rules, a majority of miners can add a new rule without creating a chain split(assuming other nodes don't change their rules). Since a majority of miners are adding a rule with BIP91 the blocks will be valid to existing nodes. There's mostly 2 arguments I'm seeing, one is that more users should enforce the BIP91 rules to strengthen the rule enforcement, the other is that users should not so that there is more consistency among rule enforcement. I somewhat agree with the consistency argument, however any businesses accepting transactions without many confirmations should be running both BIP91 and non-BIP91 nodes and ensuring that the transactions are confirmed on both.

dooglus commented:
> however any businesses accepting transactions without many confirmations should be running both BIP91 and non-BIP91 nodes and ensuring that the transactions are confirmed on both

Exactly. That seems like an argument for merging this PR so that I can run a BIP91 node without having to trust or code review a non-core repository.

achow101 commented:
Can this be squashed and possibly updated with the latest from the segsignal repo?

jameshilliard commented:
Closing this as enforcement of BIP91 is no longer required.
